pkg/certificates/renewer.go

```diff
@@ -0,0 +1,202 @@
+	if !ShouldProcessComponent(component, constants.EtcdComponent) {
+		return nil
+	}
+
+	if len(config.Etcd.Nodes) == 0 {
+		logger.Info("Cluster does not have external ETCD.")
+		return nil
+	}
+
+	logger.MarkPass("Starting etcd certificate renewal process")
+
+	if err := r.renewEtcdCerts(ctx, config); err != nil {
+		return fmt.Errorf("renewing etcd certificates: %v", err)
+	}
+
+	logger.MarkSuccess("Etcd certificate renewal process completed successfully.")
+	return nil
+}
+
+// processControlPlaneRenewal handles the renewal of control plane certificates if needed.
+func (r *Renewer) processControlPlaneRenewal(ctx context.Context, config *RenewalConfig, component string) error {
+	if !ShouldProcessComponent(component, constants.ControlPlaneComponent) {
+		return nil
+	}
+
+	if len(config.ControlPlane.Nodes) == 0 {
+		return fmt.Errorf("❌ Error: No control plane node IPs found")
+	}
+
+	logger.MarkPass("Starting control plane certificate renewal process")
+	if err := r.renewControlPlaneCerts(ctx, config, component); err != nil {
+		return fmt.Errorf("renewing control plane certificates: %v", err)
+	}
+
+	logger.MarkSuccess("Control plane certificate renewal process completed successfully.")
+	return nil
+}
+
+// finishRenewal performs cleanup operations after certificate renewal.
+func (r *Renewer) finishRenewal() error {
+	logger.MarkPass("Cleaning up temporary files")
+	if err := r.cleanup(); err != nil {
+		logger.MarkFail("API server unreachable — skipping cleanup to preserve debug data.")
+		return err
+	}
+
+	logger.MarkPass("Cleanup completed")
+	return nil
+}
+
+// RenewCertificates orchestrates the certificate renewal process.
+func (r *Renewer) RenewCertificates(ctx context.Context, _ *types.Cluster,
+	cfg *RenewalConfig, component string,
+) error {
+	if err := ValidateComponent(component); err != nil {
+		return err
+	}
+
+	// make sure if API is not reachable, stil proceed to ssh
+	if err := r.ssh.InitSSHConfig(cfg.ControlPlane.SSHUser,
+		cfg.ControlPlane.SSHKey, cfg.ControlPlane.SSHPasswd); err != nil {
+		return fmt.Errorf("init ssh: %v", err)
+	}
+
+	if err := r.kubeClient.CheckAPIServerReachability(ctx); err != nil {
+		logger.MarkWarning("API server unreachable, kubeconfig might be expired", "error", err)
+
+		// pull new kubeconfig
+		newCfg, recErr := RecoverExpiredKubeconfig(ctx, cfg, r.ssh)
+		if recErr != nil {
+			return fmt.Errorf("auto-recover kubeconfig failed: %v", recErr)
+		}
+		if err := r.kubeClient.InitClientWithKubeconfig(newCfg); err != nil {
+			return fmt.Errorf("re-init client with recovered kubeconfig: %v", err)
+		}
+
+		if err := r.kubeClient.CheckAPIServerReachability(ctx); err != nil {
+			return fmt.Errorf("API server still unreachable after kubeconfig recovery: %v", err)
+		}
+	}
+
+	if err := r.kubeClient.BackupKubeadmConfig(ctx, r.backupDir); err != nil {
+		return fmt.Errorf("backup kubeadm-config: %v", err)
+	}
+	if err := r.processEtcdRenewal(ctx, cfg, component); err != nil {
```

Comment:
Can we move validation part at one place? Same for control plane
```suggestion
	if err := r.renewEtcdCerts(...); err != nil {
```